feat!: multiple buckets / providers support (#48)

hbollon opened the pull request:
As the title shown, I've added the multiple buckets / providers support into Terraboard.
It's now possible to set multiples S3/MinIO buckets or to mix providers (to have AWS, Gitlab and GCP simultaneously for example), Terraboard will fetch them all on db refresh.

However: 
- To use this functionality, the only configuration method usable is through Yaml config file. Env and flags are still usable for mono bucket configuration (maybe we can use them for multiple providers but it's not appropriated imo)
- Atm, Terraboard is dependent of **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** env variables for AWS provider, which prevents being able to have several backend AWS or others compatible providers like MinIO (we can still have several buckets under the same backend). We could consider making these values configurable via the yaml file however (@raphink what do you think?).

I tested that with MinIO/AWS but not with others providers since I don't have any account for them.
The `terraboard-dev-multi-providers` service in the `test/` docker-compose is configured to use two MinIO bucker (just replace `terraboard-dev` by it in the compose file)
I also updated the ReadMe with some instructions / advices.

Should meets the needs of #48 :smile:

raphink commented on README.md:
```diff
@@ -89,6 +90,9 @@ It currently supports several remote state backend providers:
 - [Terraform Cloud (remote)](https://www.terraform.io/docs/backends/types/remote.html)
 - [GitLab](https://docs.gitlab.com/ee/user/infrastructure/terraform_state.html)
 
+With the upcoming **v1.2.0** update, Terraboard will be now able to handle multiple buckets/providers configuration! 🥳
```
That's a nice addition. We'll just need to remember to remove it when we release 😄

raphink replied:
```suggestion
With the upcoming **v2.0.0** update, Terraboard will be now able to handle multiple buckets/providers configuration! 🥳
```

raphink commented on README.md:
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
+2. CLI parameters **(only usable for mono provider configuration)**
+3. Configuration file (YAML). A configuration file example can be found in the root directory of this repository and in the `test/` subdirectory.
+
+**Important: all flags/environment variables related to the providers settings aren't compatible with multi-provider configuration! For that, you must use the Yaml config file to be able to configure multiples buckets/providers.**
```
```suggestion
**Important: all flags/environment variables related to the providers settings aren't compatible with multi-provider configuration! Instead, you must use the YAML config file to be able to configure multiples buckets/providers.**
```

raphink commented on README.md:
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
 
+### Multiple buckets/providers
+
+In order to be able to link to Terraboard multiples buckets or even providers, you must use the Yaml configuration method.
```
```suggestion
In order for Terraboard to import states from multiples buckets or even providers, you must use the YAML configuration method.
```

raphink commented on README.md:
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
+
+In order to be able to link to Terraboard multiples buckets or even providers, you must use the Yaml configuration method.
+
+- Set the environment variable **CONFIG_FILE** or the flag **-c** / **--config-file** to a valid Yaml config file.
```
```suggestion
- Set the environment variable **CONFIG_FILE** or the flag **-c** / **--config-file** to point to a valid YAML config file.
```

raphink commented on README.md:
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
+
+- Set the environment variable **CONFIG_FILE** or the flag **-c** / **--config-file** to a valid Yaml config file.
+
+- In the Yaml file, specify your desired providers configuration. For example with two S3 buckets:
```
```suggestion
- In the YAML file, specify your desired providers configuration. For example for two S3 buckets:
```

raphink replied:
Don't we need to specify the no versioning and no locks options in this example since it uses minio?

raphink commented on README.md:
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
+      force-path-style: true
+      file-extension: 
+        - .tfstate
+
```
```suggestion
```

raphink commented on README.md:
````diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
+
+```
+
+In the case of AWS, don't forget to set **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** environment variables.
````
```suggestion
In the case of AWS, don't forget to set the **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** environment variables.
```

raphink commented on README.md:
```diff
@@ -142,12 +146,47 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
+In the case of AWS, don't forget to set **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** environment variables.
+
+That's it! Terraboard will now fetch these two buckets on DB refresh. You can also mix providers like AWS and Gitlab or anything else.
+You can find a ready-to-use docker example with two *MinIO* buckets in the `test/` sub-folder (just swipe the two terraboard services in the docker-compose file). 
```
```suggestion
You can find a ready-to-use Docker example with two *MinIO* buckets in the `test/` sub-folder (just swipe the two terraboard services in the docker-compose file). 
```

raphink replied:
What do you mean by "swipe the services"?

hbollon replied:
> What do you mean by "swipe the services"?

I added in the docker-compose of `/test`, a Terraboard container configured to use a YAML config file and two MinIO buckets.
However, I commented it in favor of the Terraboard lambda container to avoid having two applications simultaneously.

```yaml
services:
  terraboard-dev:
    build:
      context: ../
      dockerfile: ./Dockerfile
    environment:
      AWS_ACCESS_KEY_ID: root
      AWS_SECRET_ACCESS_KEY: mypassword
      AWS_BUCKET: test-bucket
      AWS_REGION: eu-west-1
      AWS_ENDPOINT: http://minio:9000/
      AWS_FORCE_PATH_STYLE: "true"
      TERRABOARD_LOG_LEVEL: debug
      TERRABOARD_NO_LOCKS: "true"
      TERRABOARD_NO_VERSIONING: "true"
      DB_PASSWORD: mypassword
      DB_SSLMODE: disable
      GODEBUG: netdns=go
    depends_on:
      db:
        condition: service_healthy
      minio:
        condition: service_started
    volumes:
      - ../static:/static:ro
    ports:
      - "8080:8080"

  # terraboard-dev-multi-providers:
  #   build:
  #     context: ../
  #     dockerfile: ./Dockerfile
  #   environment:
  #     AWS_ACCESS_KEY_ID: root
  #     AWS_SECRET_ACCESS_KEY: mypassword
  #     DB_SSLMODE: disable
  #     CONFIG_FILE: config/config.yml
  #     GODEBUG: netdns=go
  #   depends_on:
  #     db:
  #       condition: service_healthy
  #     minio:
  #       condition: service_started
  #   volumes:
  #     - ../static:/static:ro
  #     - ./config.yml:/config/config.yml:ro
  #   ports:
  #     - "8081:8080"
```

I'm not really happy with that, maybe we can split the test sub-folder into multiple config folders each with a docker-compose :thinking:

raphink replied:
Ah, so you mean "swap" then, not "swipe" :-)

hbollon replied:
Yes :laughing:

hbollon replied:
Yes :laughing:

raphink commented on api/api.go:
```diff
@@ -147,15 +147,21 @@ func StateCompare(w http.ResponseWriter, r *http.Request, d *db.Database) {
-	if err != nil {
-		JSONError(w, "Failed to get locks", err)
-		return
+	allLocks := make(map[string]state.LockInfo)
```
What is the point of using a map rather than an array?

raphink commented on config/config.go:
```diff
@@ -119,14 +119,44 @@ func (c *Config) LoadConfigFromYaml() *Config {
+
+// LoadConfig loads the config from flags & environment
+func LoadConfig(version string) *Config {
+	c := initDefaultConfig()
+	parseStructFlagsAndEnv(&c)
```
Why is this necessary?

hbollon replied:
It's for keeping the env / flags configuration possible with single provider Terraboard.
Indeed, now we have providers array in the `config` struct:
```go
AWS []AWSConfig `group:"AWS Options" yaml:"aws"`

TFE []TFEConfig `group:"Terraform Enterprise Options" yaml:"tfe"`

GCP []GCPConfig `group:"Google Cloud Platform Options" yaml:"gcp"`

Gitlab []GitlabConfig `group:"GitLab Options" yaml:"gitlab"`
```

And so they don't have any elements in them on `Config` instanciation which prevent from using env variables or flags.
`c := initDefaultConfig()` add an empty struct in each of them and parse go-flags stuff in them. If they're still empty or invalid they will be ignored but with that process we can still use Terraboard with others configuration methods as before (although they are incompatible with multi buckets).

`parseStructFlagsAndEnv(interface{})` is just a DRY function for go-flags parsing

raphink commented on main.go:
```diff
@@ -64,56 +64,61 @@ func isKnownStateVersion(statesVersions map[string][]string, versionID, path str
-			continue
-		}
+		log.Debugf("Total providers: %d\n", len(sps))
+		for i, sp := range sps {
```
Looping inside `refreshDB()` means each provider will be parsed one a time. Maybe we could parallelize providers instead by looping ouside the call to `refreshDB()` in `main()` and keep `refreshDB()` the way it is.

hbollon replied:
Great idea that I should have thought of! 
This will allow to not interrupt all the fetching processes if one of them has an error and reduce the time that the refresh takes
I will do that

hbollon commented:
I've done the following changes: 
- RefreshDB is now parallelized 
- I added mutex lock security on sensible db operations (FirstOrCreate principally, which lead to errors without locks)
- I split test directory into multiples sub-folder containing different configurations 
- I fixed all ReadMe typo issues

raphink commented on README.md:
```diff
@@ -142,12 +146,51 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
 
+### Multiple buckets/providers
+
+In order for Terraboard to import states from multiples buckets or even providers, you must use the YAML configuration method.
+
+- Set the environment variable **CONFIG_FILE** or the flag **-c** / **--config-file** to point to a valid YAML config file.
+
+- In the YAML file, specify your desired providers configuration. For example with two MinIO buckets (using the AWS provider with compatible mode):
```
```suggestion
In order for Terraboard to import states from multiples buckets or even providers, you must use the YAML configuration method:

- Set the `CONFIG_FILE` environment variable or the `-c`/`--config-file` flag to point to a valid YAML config file.
- In the YAML file, specify your desired providers configuration. For example with two MinIO buckets (using the AWS provider with compatible mode):
```

raphink commented on README.md:
````diff
@@ -142,12 +146,51 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
+        - .tfstate
+```
+
+In the case of AWS, don't forget to set the **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** environment variables.
````
I'd use tt instead of bold for code
```suggestion
In the case of AWS, don't forget to set the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` environment variables.
```

raphink commented on db/db.go:
```diff
@@ -79,7 +81,7 @@ func Init(config config.DBConfig, debug bool) *Database {
 	}
 
-	d := &Database{db}
+	d := &Database{DB: db}
```
Not sure this change is necessary (though it doesn't hurt)

hbollon replied:
I had _InvalidStructLit_ error from gopls without it since I added `lock sync.Mutex` field

raphink commented on db/db.go:
```diff
@@ -124,11 +126,13 @@ func (db *Database) stateS3toDB(sf *statefile.File, path string, versionID strin
 	err = db.FirstOrCreate(&lineage, types.Lineage{Value: sf.Lineage}).Error
 	if err != nil || lineage.ID == 0 {
-		log.Error("Unknown error in stateS3toDB during lineage finding")
+		log.Errorf("Unknown error in stateS3toDB during lineage finding: %v", err)
```
It could be interesting to use a structured log here (with fields).

hbollon replied:
Done :+1:

Shocktrooper commented:
> As the title shown, I've added the multiple buckets / providers support into Terraboard.
> It's now possible to set multiples S3/MinIO buckets or to mix providers (to have AWS, Gitlab and GCP simultaneously for example), Terraboard will fetch them all on db refresh.
> 
> However:
> 
> * To use this functionality, the only configuration method usable is through Yaml config file. Env and flags are still usable for mono bucket configuration (maybe we can use them for multiple providers but it's not appropriated imo)
> * Atm, Terraboard is dependent of **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** env variables for AWS provider, which prevents being able to have several backend AWS or others compatible providers like MinIO (we can still have several buckets under the same backend). We could consider making these values configurable via the yaml file however (@raphink what do you think?).
> 
> I tested that with MinIO/AWS but not with others providers since I don't have any account for them.
> The `terraboard-dev-multi-providers` service in the `test/` docker-compose is configured to use two MinIO bucker (just replace `terraboard-dev` by it in the compose file)
> I also updated the ReadMe with some instructions / advices.
> 
> Should meets the needs of #48 😄

With **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** being needed does this prevent multiple cross account S3 buckets from being selected? Would they all have to be in the same account and is there work to migrate over to the AWS SDK or something that can automatically handle credentials automatically instead of relying on hardcoded environment variables

hbollon commented:
> With **AWS_ACCESS_KEY_ID** and **AWS_SECRET_ACCESS_KEY** being needed does this prevent multiple cross account S3 buckets from being selected?

Actually, with this PR, yes all AWS buckets must be on the same account.

> is there work to migrate over to the AWS SDK or something that can automatically handle credentials automatically instead of relying on hardcoded environment variables

This is exactly what I proposed to @raphink 😃

hbollon commented:
I've added:
- AWS access key and secret access key config fields and aws sdk appropriate usage to support multiple AWS accounts configuration
- AWS Session Token support (should do the trick for #181)
- The ability to directly configure multiple buckets for the same AWS account under the "S3" field of the YAML file which is now an objects array

I also updated unit test and docker testing env to match with new config (I also add a second minio container in the multiple-minio-buckets test env)
Like previously, I made sure that the config via environment variables or by flags is always possible for the instances of Terraboard with a single provider

hbollon commented on config/yaml.go:
```diff
@@ -0,0 +1,33 @@
+ * (and so makes them optional)
+ *********************************************/
+
+func (s *S3BucketConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type rawS3BucketConfig S3BucketConfig
+	raw := rawS3BucketConfig{
+		FileExtension: []string{".tfstate"},
+	}
+	if err := unmarshal(&raw); err != nil {
+		return err
+	}
+
+	*s = S3BucketConfig(raw)
+	return nil
+}
+
+func (s *GitlabConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	type rawGitlabConfig GitlabConfig
+	raw := rawGitlabConfig{
+		Address: "https://gitlab.com",
+	}
+	if err := unmarshal(&raw); err != nil {
+		return err
+	}
+
+	*s = GitlabConfig(raw)
+	return nil
+}
```
Yes, without that we don't have default values for Yaml config method

hbollon commented on state/aws.go:
```diff
@@ -31,39 +31,45 @@ type AWS struct {
-	sess := session.Must(session.NewSession())
-
-	awsConfig := aws_sdk.NewConfig()
+func NewAWS(c *config.Config) []*AWS {
```
To avoid the iteration in `NewAWS()`?

raphink replied:
Yes, and also because generally I'd expect an instantiation method to return a single instance.

hbollon replied:
Ok so I will do that for AWS but also for all others providers
Each one will have a `NewProvider()` function wich return a single instance and a `NewProviderCollection()` one which will iterate and call the other one. (ofc in functions names _**Provider**_ will be replaced by his name)

Shocktrooper commented:
Is there since there was mention of aws sdk usage of credentials is there an example of how to tell terraboard to assume a role in another account or how to tell it to use an instance profile in the case of S3 buckets

raphink commented on README.md:
```diff
@@ -142,12 +146,50 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
-1. Environment variables
-2. CLI parameters
-3. Configuration file (YAML). A configuration file example can be found in the root directory of this repository.
+1. Environment variables **(only usable for mono provider configuration)**
```
```suggestion
1. Environment variables **(only usable for single provider configuration)**
```

raphink commented on README.md:
```diff
@@ -142,12 +146,50 @@ Data resiliency is not paramount though as this dataset can be rebuilt upon your
-2. CLI parameters
-3. Configuration file (YAML). A configuration file example can be found in the root directory of this repository.
+1. Environment variables **(only usable for mono provider configuration)**
+2. CLI parameters **(only usable for mono provider configuration)**
```
```suggestion
2. CLI parameters **(only usable for single provider configuration)**
```

raphink commented on state/aws.go:
```diff
@@ -31,39 +32,59 @@ type AWS struct {
-			if c.AWS.ExternalID != "" {
-				p.ExternalID = aws_sdk.String(c.AWS.ExternalID)
+func NewAWS(aws config.AWSConfig, bucket config.S3BucketConfig, noLocks, noVersioning bool) *AWS {
+	if bucket.Bucket != "" {
```
Could you shortcut this case instead?

```go
if bucket.Bucket == "" {
  return nil
}
```

raphink commented on state/gcp.go:
```diff
@@ -24,33 +24,53 @@ type GCP struct {
+	var gcpInstance *GCP
 	var err error
-	if c.GCP.GCPSAKey != "" {
+	if gcp.GCSBuckets != nil {
```
Same here, I think the bucket logic could be short circuited

raphink commented on state/gitlab.go:
```diff
@@ -18,10 +18,25 @@ type Gitlab struct {
-		Client: gitlab.NewClient(c.Gitlab.Address, c.Gitlab.Token),
+func NewGitlab(gl config.GitlabConfig) *Gitlab {
+	var instance *Gitlab
+	if gl.Token != "" {
```
Same here, short circuit the empty token

raphink commented on state/state.go:
```diff
@@ -40,23 +40,53 @@ type Provider interface {
+	if len(c.TFE) > 0 {
+		objs, err := NewTFECollection(c)
+		if err != nil {
+			return []Provider{}, err
```
If you named the return values of the method, you might just use `return` here iirc

raphink replied:
Note, if one provider fails, you get no providers at all at the moment.

hbollon replied:
> Note, if one provider fails, you get no providers at all at the moment.
Do you think it's better to just log an error and continue with the following ones?
Initially, I considered this error as fatal since it would lead to the non-inclusion of an entire provider and because, potentially, it could come from an error in the configuration file which could impact the rest.

raphink replied:
Not sure. I say I'd prefer to see some data and get an error somewhere rather than nothing at all.

hbollon replied:
> If you named the return values of the method, you might just use return here iirc

Indeed but in this case I think it's more simple like this since if I name the return values I can't use `:=` anymore with err which will force me to declare the temp variable `objs` before each time I use it. It's, imo, more complex and furthermore it raise an error on `gosimple` if I do it.

mcanevet reviewed (changes requested):
I guess you should change one of your commits' commit message with an exclamation mark as it contains a breaking change.

raphink commented on state/tfe.go:
```diff
@@ -19,24 +19,42 @@ type TFE struct {
-	}
+func NewTFE(tfeObj config.TFEConfig) (*TFE, error) {
+	var tfeInstance *TFE
+	if tfeObj.Token != "" {
```
Short circuit
